File: design-pattern/iterator-pattern/iterator.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>

using namespace std;

template <typename T>
class Iterator
{
public:
    virtual bool hasNext() = 0;
    virtual T next() = 0;
    virtual ~Iterator() {}
};

template <typename T>
class Iterable
{
public:
    virtual Iterator<T> *getIterator() = 0;
};

class LinkList : public Iterable<int>
{
public:
    int data;
    LinkList *next;

    LinkList(int val)
    {
        data = val;
        next = nullptr;
    }

    Iterator<int> *getIterator() override;
};

class BinearyTree : public Iterable<int>
{
public:
    int data;
    BinearyTree *left;
    BinearyTree *right;

    BinearyTree(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }

    Iterator<int> *getIterator() override;
};

class Song
{
public:
    string name;
    string artist;

    Song(string n, string a)
    {
        name = n;
        artist = a;
    }
};

class Playlist : public Iterable<Song *>
{
public:
    vector<Song *> items;

    void addSong(Song *s)
    {
        items.push_back(s);
    }

    Iterator<Song *> *getIterator() override;
};

class LinkListIterator : public Iterator<int>
{
private:
    LinkList *curr;

public:
    LinkListIterator(LinkList *head)
    {
        curr = head;
    }

    bool hasNext() override
    {
        return curr != nullptr;
    }

    int next() override
    {
        int val = curr->data;
        curr = curr->next;
        return val;
    }
};

class BinearyTreeIterator : public Iterator<int>
{
private:
    stack<BinearyTree *> st;

    void pushLeft(BinearyTree *node)
    {
        while (node)
        {
            st.push(node);
            node = node->left;
        }
    }

public:
    BinearyTreeIterator(BinearyTree *root)
    {
        pushLeft(root);
    }

    bool hasNext() override
    {
        return !st.empty();
    }

    int next() override
    {
        BinearyTree *node = st.top();
        st.pop();

        int val = node->data;

        if (node->right)
        {
            pushLeft(node->right);
        }

        return val;
    }
};

/* ===================== Playlist Iterator ===================== */
class PlayListIterator : public Iterator<Song *>
{
private:
    vector<Song *> &songs;
    size_t index;

public:
    PlayListIterator(vector<Song *> &s)
        : songs(s), index(0) {}

    bool hasNext() override
    {
        return index < songs.size();
    }

    Song *next() override
    {
        return songs[index++];
    }
};

Iterator<int> *LinkList::getIterator()
{
    return new LinkListIterator(this);
}

Iterator<int> *BinearyTree::getIterator()
{
    return new BinearyTreeIterator(this);
}

Iterator<Song *> *Playlist::getIterator()
{
    return new PlayListIterator(items);
}

int main()
{
    /* Linked List */
    LinkList *l1 = new LinkList(1);
    l1->next = new LinkList(2);
    l1->next->next = new LinkList(3);

    Iterator<int> *listIt = l1->getIterator();
    while (listIt->hasNext())
    {
        cout << listIt->next() << " ";
    }
    cout << endl;

    /* Binary Tree */
    BinearyTree *tree = new BinearyTree(1);
    tree->left = new BinearyTree(2);
    tree->right = new BinearyTree(3);

    Iterator<int> *treeIt = tree->getIterator();
    while (treeIt->hasNext())
    {
        cout << treeIt->next() << " ";
    }
    cout << endl;

    /* Playlist */
    Playlist *pl = new Playlist();
    pl->addSong(new Song("Song1", "Artist1"));
    pl->addSong(new Song("Song2", "Artist2"));

    Iterator<Song *> *playIt = pl->getIterator();
    while (playIt->hasNext())
    {
        Song *s = playIt->next();
        cout << "Title: " << s->name
             << ", Artist: " << s->artist << endl;
    }

    return 0;
}

File: design-pattern/iterator-pattern/iterator.ts
interface Iterator<T> {
    hasNext(): boolean;
    next(): T;
}

interface Iterable<T> {
    getIterator(): Iterator<T>;
}

// ---------------- Linked List ----------------
class LinkList implements Iterable<number> {
    public next: LinkList | null;
    public data: number;

    constructor(data: number, next: LinkList | null = null) {
        this.data = data;
        this.next = next;
    }

    getIterator(): Iterator<number> {
        return new LinkListIterator(this);
    }
}

class LinkListIterator implements Iterator<number> {
    private head: LinkList | null;

    constructor(curr: LinkList) {
        this.head = curr;
    }

    hasNext(): boolean {
        return this.head !== null;
    }

    next(): number {
        if (!this.head) {
            throw new Error("No more elements");
        }

        const val = this.head.data;
        this.head = this.head.next;

        return val;
    }
}

// ---------------- Binary Tree ----------------
class BinaryTree implements Iterable<number> {
    public left: BinaryTree | null;
    public right: BinaryTree | null;
    public data: number;

    constructor(data: number) {
        this.data = data;
        this.left = null;
        this.right = null;
    }

    getIterator(): Iterator<number> {
        return new BinaryTreeIterator(this);
    }
}

class BinaryTreeIterator implements Iterator<number> {
    private stack: BinaryTree[] = [];

    constructor(root: BinaryTree) {
        this.pushLeft(root);
    }

    private pushLeft(node: BinaryTree | null): void {
        while (node) {
            this.stack.push(node);
            node = node.left;
        }
    }

    hasNext(): boolean {
        return this.stack.length > 0;
    }

    next(): number {
        if (!this.hasNext()) {
            throw new Error("No more elements");
        }

        const node = this.stack.pop()!;

        if (node.right) {
            this.pushLeft(node.right);
        }

        return node.data;
    }
}

// ---------------- Playlist ----------------
class Song {
    constructor(
        public title: string,
        public artist: string
    ) { }
}

class PlayList implements Iterable<Song> {
    private songs: Song[] = [];

    addSong(song: Song): void {
        this.songs.push(song);
    }

    getIterator(): Iterator<Song> {
        return new PlayListIterator(this.songs);
    }
}

class PlayListIterator implements Iterator<Song> {
    private index: number = 0;

    constructor(private songs: Song[]) { }

    hasNext(): boolean {
        return this.index < this.songs.length;
    }

    next(): Song {
        if (!this.hasNext()) {
            throw new Error("No more elements");
        }

        return this.songs[this.index++];
    }
}

// ---------------- Client Code ----------------
const list = new LinkList(1);
list.next = new LinkList(2);
list.next.next = new LinkList(3);

console.log("Linked List:");
const listIterator = list.getIterator();

while (listIterator.hasNext()) {
    console.log(listIterator.next());
}

const tree = new BinaryTree(1);
tree.left = new BinaryTree(2);
tree.right = new BinaryTree(3);

console.log("\nBinary Tree:");
const treeIterator = tree.getIterator();

while (treeIterator.hasNext()) {
    console.log(treeIterator.next());
}

const playlist = new PlayList();

playlist.addSong(new Song("Song1", "Artist1"));
playlist.addSong(new Song("Song2", "Artist2"));

console.log("\nPlaylist:");
const playlistIterator = playlist.getIterator();

while (playlistIterator.hasNext()) {
    const song = playlistIterator.next();
    console.log(`${song.title} - ${song.artist}`);
}